Client/Application/AppDelegate.swift

```diff
@@ -96,6 +108,36 @@ class AppDelegate: UIResponder, UIApplicationDelegate {
                     }
                 }
             }
+        } else if let components = NSURLComponents(URL: url, resolvingAgainstBaseURL: false) where components.scheme == "firefox" && components.host == "x-callback-url" {
+            if let query = components.query, items = components.queryItems as? [NSURLQueryItem] {
+                if items[0].name == "x-source" && items[0].value != nil && items[1].name == "url" && items[1].value != nil {
+                    if let newURL = NSURL(string: items[1].value!) {
+                        let tab = self.tabManager.addTab(request: NSURLRequest(URL: newURL))
+                        self.tabManager.selectTab(tab)
+                        if let backToAppURL = NSURL(string: items[0].value!) {
+                            if let appName = items[2].value where items[2].name == "x-source-name" {
+                                if UIApplication.sharedApplication().canOpenURL(backToAppURL) {
+
+                                    //button code will change when we use the button with image asset
+                                    var button = UIButton()
+                                    button.setTitle("〈 \(appName)", forState: UIControlState.Normal)
+                                    var fontSize: CGFloat = 14
+                                    button.titleLabel!.font = UIFont(name: "HelveticaNeueInterface", size: fontSize)
+                                    button.setTitleColor(UIColor.grayColor(), forState: UIControlState.Normal)
+                                    button.titleLabel!.lineBreakMode = NSLineBreakMode.ByTruncatingTail
+                                    button.addTarget(self, action: "backToApp", forControlEvents: UIControlEvents.TouchUpInside)
+
+                                    tab.callbackToAppButton = button
+                                    tab.callbackURL = backToAppURL
+                                    self.browserViewController.urlBar.addSubview(button)
+                                    self.browserViewController.urlBar.updateConstraints()
```

Comment:
This is kinda wrong. You're directly manipulating the `tabManager` above, and you're manipulating the controller here.

What we really want is either:

* Make the tab, give it the attribute, and then switch to it — the right UI elements will appear as a result of switching
* Ask the BVC to create the tab with the right attribute and display.

You're splitting the difference between the two, which means there's code in this AppDelegate that constructs a button, but you also need to understand and touch BVC.